node: sample-device-plugin: register to kubelet by default and ensure re-registration to kubelet on kubelet restarts

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

In #115107 we added an environment variable to control the registration of sample
device plugin to kubelet. The intent of this patch is to ensure that the default
behaviour of the plugin is to register to kubelet (in case no environment
variable is specified).
In addition to that, we want to ensure that the plugin registers itself not just once.
It should re-register itself to kubelet in case of node reboot or kubelet restarts.

#### Which issue(s) this PR fixes:
Related to: 

-  https://github.com/kubernetes/kubernetes/pull/115107 
-  https://github.com/kubernetes/kubernetes/pull/116376


#### Special notes for your reviewer:
This PR would require a follow-up to promote the sample device plugin image similar to the work done [here](https://github.com/kubernetes/k8s.io/pull/4862).

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```
<!--
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>

```docs

```
-->

swatisehgal committed:
node: sample-dp: register by default and re-register on restarts

In issue: 115107 we added an environment variable to control the registration of sample
device plugin to kubelet. The intent of this patch is to ensure that the default
behaviour of the plugin is to register to kubelet (in case no environment
variable is specified).

In addition to that, we want to ensure that the plugin registers itself not just once.
It should re-register itself to kubelet in case of node reboot or kubelet restarts.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/sig node

swatisehgal commented:
/cc @ffromani

swatisehgal commented:
/retest

swatisehgal commented:
The files and commit is not rendering properly for in the PR for reason.
Here is the diff with the same branch with which the PR was created: https://github.com/kubernetes/kubernetes/compare/master...swatisehgal:kubernetes:sample-dp-register-by-default.

swatisehgal commented:
> The files and commit are not rendering properly for in the PR for reason. Here is the diff with the same branch with which the PR was created: [master...swatisehgal:kubernetes:sample-dp-register-by-default](https://github.com/kubernetes/kubernetes/compare/master...swatisehgal:kubernetes:sample-dp-register-by-default).

Looks like I ran into some git error. Pushed a dummy commit which I  reverted later and that seemed to have helped.

SergeyKanzhelev commented:
/assign

SergeyKanzhelev commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -94,17 +94,26 @@ func main() {
+		// if REGISTER_CONTROL_FILE environment variable is not explicitly specified, we would like to register the device
+		// plugin to kubelet as normal. The registration is added in a for loop to allow re-registration of the
+		// plugin to kubelet in case of node reboot or kubelet restarts.
+		for {
```
do you need some sort of wait in-between of registrations? Otherwise it will be consuming a lot of CPU

swatisehgal replied:
Sure, going to work on improvements.

ffromani commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -94,17 +94,26 @@ func main() {
+			if err := dp1.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
+				panic(err)
+			}
+		}
```
I fully agree with the intent of this PR. Got burned already by the plugin not reconnecting. Designwise full ack. Implementationwise, let's explore these two points:

1. I concur a registration busy loop is probably not the best approach, unless `Register` is blocking and the pluin can register just once (is this documented? I don't recall). Would be better to register, sleep, detect disconnection, attempt to reconnect until succeed
2. we need to clarify and make explicit if REGISTER_CONTROL_FILE controls _A_ registration or ALL registration attempts. IOW, should the plugin optionally block _every_ connection attempt including reconnections or just the first? All options are on the table, let's decide and document.

swatisehgal replied:
> I fully agree with the intent of this PR. Got burned already by the plugin not reconnecting. Designwise full ack. Implementationwise, let's explore these two points:
> 
> 1. I concur a registration busy loop is probably not the best approach, unless `Register` is blocking and the pluin can register just once (is this documented? I don't recall). Would be better to register, sleep, detect disconnection, attempt to reconnect until succeed

Sure, wanted to have something quick and dirty to verify if the kubelet restart issue reported [here](https://github.com/kubernetes/kubernetes/issues/118559) is real or passed in the e2e tests due to the sample device plugin implementation. In addition to that, I wanted to get a general consensus on whether this is something we want to fix. Looks like we have that now, so let's totally improve the implementation.  

  > 2. we need to clarify and make explicit if REGISTER_CONTROL_FILE controls _A_ registration or ALL registration attempts. IOW, should the plugin optionally block _every_ connection attempt including reconnections or just the first? All options are on the table, let's decide and document.

With the introduction of the `REGISTER_CONTROL_FILE` the default behavior of automatic registration was changed. In the absence of control file or environment variable, we now don't register which might be unexpected and I was attempting to rectify that as part of this PR as well which might explain the rationale behind the title.

In my opinion we should have two flows:
1. Case 1: Where REGISTER_CONTROL_FILE is not specified or is empty: In this case we follow the normal flow of registering to kubelet by default. This is applicable for all the registrations of the plugin to kubelet.
2. Case 2: If the environment variable is specified with a control file, the deletion of the control file triggers registration. Again, this should happen for ALL the registrations when environment variable is specified.

swatisehgal replied:
> Additionally, we need to review the expectations for device plugins. Are they expected to reconnect? [EDIT] [yes, they are.](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#handling-kubelet-restarts)[/EDIT]

Thanks for checking this.

swatisehgal committed:
node: sample-dp: Handle re-registration for controlled registrations

In case `REGISTER_CONTROL_FILE` is specified, we want to ensure that the
registration is triggered by deletion of the control file. This is
applicable both when the registration happens for the first time and
subsequent ones because of kubelet restarts.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -184,7 +184,7 @@ func (m *Stub) Stop() error {
 }
 
-func (m *Stub) Watch(resourceName string) {
+func (m *Stub) Watch(resourceName string, autoregister bool) {
```
iff we need to add this feature to the `stub`, I think it should be a global option. Perhaps we should add a configurable callback like `stubGetPreferredAllocFunc` to control the registration?
I believe it should _not_ be a Watch option because is supposed to always hold the same value for the lifecycle of the plugin. IOW, we don't expect this function to be called sometimes with true, sometimes with false. Should always be consistent.

swatisehgal replied:
> iff we need to add this feature to the stub, I think it should be a global option. Perhaps we should add a configurable callback like stubGetPreferredAllocFunc to control the registration?

Are you referring to the use of `GetPreferredAllocation` device plugin endpoint here or is the name co-incidental?  IIUC, that allows device plugin to provide its own preference for device allocation. I am not sure how we use that to control registration.  

> I believe it should not be a Watch option because is supposed to always hold the same value for the lifecycle of the plugin. IOW, we don't expect this function to be called sometimes with true, sometimes with false. Should always be consistent.

I am not sure I fully understand this comment. The overall intent of watch here is to watch for changes to the unix socket as kubelet deletes them when it starts. Once the kubelet restart is detected,  the gRPC should always be restarted. What we are doing with the `autoregister` argument is facilitating the cases where the device plugin wants to take control of its own re-registration  (which it can do by  specifying autoregister= false and happens in case of sample device plugin when it has REGISTER_CONTROL_FILE specified as environment var). Typical flow would involve both start of gRPC server and registration of the plugin to kubelet (autoregister= true).

ffromani replied:
Sorry for my extremely terse and hard to understand comments. Let me elaborate a bit

>Are you referring to the use of GetPreferredAllocation device plugin endpoint here or is the name co-incidental? IIUC, that allows device plugin to provide its own preference for device allocation. I don't think that can be used to control the registration process.

I'd experiment adding a callback which client code can use as hook point to control the registration process. It's probably easier for all of us if I sketch out some code. Something like this:

```
diff --git a/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go b/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
index 1d226f72a85..8d0bd4b2f01 100644
--- a/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
+++ b/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
@@ -53,11 +53,20 @@ type Stub struct {
 	// getPreferredAllocFunc is used for handling getPreferredAllocation request
 	getPreferredAllocFunc stubGetPreferredAllocFunc
 
+	// registerControlFunc is used to gate the registration requests
+	registerControlFunc stubRegisterControlFunc
+
 	registrationStatus chan watcherapi.RegistrationStatus // for testing
 	endpoint           string                             // for testing
 
 }
 
+type stubRegisterControlFunc func(err error) bool
+
+func defaultRegisterControlFunc(err error) bool {
+	return true
+}
+
 // stubGetPreferredAllocFunc is the function called when a getPreferredAllocation request is received from Kubelet
 type stubGetPreferredAllocFunc func(r *pluginapi.PreferredAllocationRequest, devs map[string]pluginapi.Device) (*pluginapi.PreferredAllocationResponse, error)
 
@@ -90,6 +99,7 @@ func NewDevicePluginStub(devs []*pluginapi.Device, socket string, name string, p
 
 		allocFunc:             defaultAllocFunc,
 		getPreferredAllocFunc: defaultGetPreferredAllocFunc,
+		registerControlFunc:   defaultRegisterControlFunc,
 	}
 }
 
@@ -103,6 +113,44 @@ func (m *Stub) SetAllocFunc(f stubAllocFunc) {
 	m.allocFunc = f
 }
 
+// SetRegisterControlFunc sets registerControlFunc of the device plugin
+func (m *Stub) SetRegisterControlFunc(f registerControlFunc) {
+	m.registerControlFunc = f
+}
+
+func (m *Stub) Watch(resourceName string) {
+	for {
+		select {
+		case stop := <-m.stopWatcher:
+			if stop {
+				return
+			}
+		default:
+			_, statErr := os.Lstat(m.socket)
+			if statErr != nil {
+				// Socket file not found; restart server
+				klog.InfoS("Server endpoint not found", m.socket)
+				klog.InfoS("Most likely Kubelet restarted")
+				if err := m.Restart(); err != nil {
+					klog.ErrorS(err, "Unable to restart server")
+					panic(err)
+
+				}
+
+				// we should probably loop until the callback passes, or clarify the semantic
+				// from stub perspective of restarting without reconnecting
+				if ok := m.registerControlFunc(statErr); ok {
+					if err := m.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
+						klog.ErrorS(err, "Unable to register to kubelet")
+						panic(err)
+					}
+				}
+			}
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
 // Start starts the gRPC server of the device plugin. Can only
 // be called once.
 func (m *Stub) Start() error {
```

The reason I'm proposing a new field in `Stub` object here is because, much like other callbacks, I expect the registration behavior to be consistent across calls of `Watch`. IOW, `Watch` should likely be called with always the same set of parameters, invariant wrt the calls. This hints the property should be aobject property, not a call parameter. Furthermore, this enables us to set defaults and enable consumers of this library which are fine with the default behavior (always register) to safely ignore this details.

Arguably, we don't really need a callback. We may use a simple object field like a boolean. Using a callback seems however more consistent with the rest of the code and enables the client code to inject delays for free, so it seems worth the little extra complexity.

swatisehgal replied:
Thanks for the suggestion, I see your point now. I will play around with the code and see how I can factor in your suggestion.

ffromani replied:
Absolutely. I'm totally fine with partially adopting my suggestion above. My main goal is to try to enable good defaults and don't necessarily expose client code to detail which may be irrelevant in some cases (e.g. ALWAYS requiring a parameter)

ffromani commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -144,13 +149,33 @@ func (m *Stub) Start() error {
 	return nil
 }
 
+func (m *Stub) Restart() error {
```
The work you're doing here looks nice at glance, but I'm a bit worried that this very-test specific behavior (which has no other known usecase outside testing, has it?) requires changes to the common stub code. Is it possible to have the same functionality, or a good enough approximation, only in application code?

ffromani replied:
Good point. What I'm concerned about is that I'm not aware of obvious use case for this feature outside the sample device plugin, but I concur this is generic enough to be a worthy addition, and emphasize the requirement of handling restarts. Let's go with it.

ffromani commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -144,13 +149,33 @@ func (m *Stub) Start() error {
+	m.server = nil
+
+	m.Start()
+	return m.Start()
```
(I probably commented already, sorry for the repetition) why call `m.Start()` twice?

swatisehgal replied:
Might have been a copy paste error, will fix in the next upload. Thanks for pointing this out.

ffromani commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -144,13 +149,33 @@ func (m *Stub) Start() error {
 	}
+	// Send terminate signal to ListAndWatch()
+	m.termSignal <- true
+	m.stopWatcher <- true
```
can we extract helpers from `Start` and `Stop` and reimplement `Restart` composing these helpers or there are complications which prevent or make this approach unpractical?

swatisehgal replied:
Ack, I will play around with that, we could potentially prevent some repetition of code by using helpers.

swatisehgal replied:
I tried creating a helper that it can be called in both `Restart` and `Stop` but that caused unit tests failures so decided to keep the code as is for now. I think the unit tests are relying on certain behavior of `Stop` and `Start` methods so changing them is causing some issues. If you feel strongly about this, I can do to deep dive into the specific unit tests and invest more time into this as it is not as straightforward as I thought.

swatisehgal commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -159,6 +184,37 @@ func (m *Stub) Stop() error {
+
+				}
+				if autoregister {
+					if err := m.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
```
The only varying component here is the `resourceName` which would depend on the device plugin. Other than that we need kubelet socket and the device plugin path which should be consistent for all registrations and plugins. 

Maybe I am not able to fully understand your suggestion and a re-phrase might help. Thanks.

ffromani replied:
Sure thing.

So, the expected flow is the client code calls `Register` and then issues `Watch` to make the stub library code handle reconnection automatically, right?
If so, ultimately, `Watch` can issue `Register` again on client code's behalf.
If so, we need to either enable the client code to pass the same parameters used by `Register` to `Watch`, or `Watch` handle this automatically.

Now, Register has these parameters `Register(kubeletEndpoint, resourceName string, pluginSockDir string)`
While Watch calls it like this `Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath)`
it seems to me that `Watch` effectively hardcodes some parameters, while the client code can use different and custom values for `kubeletEndPoint` and `pluginSockDir`. Is this the case?

swatisehgal replied:
Thanks for elaborating and for being so patient. I see your point, the parameters passed by the client to `Register` and `Watch` (which can further call register) should be the same and we should facilitate that. I will update the code taking this into consideration.

ffromani commented:
/triage accepted
/priority important-soon

it seems we may need this PR as prerequisite for https://github.com/kubernetes/kubernetes/pull/118635
I'll revisit further the priority if confirmed

swatisehgal committed:
node: sample-dp: stubRegisterControlFunc for controlling registration

If the user specifies the intent to control registration process, we rely on
registration triggers (deletion of control file) to prompt registration.

This behvaiour is expected to be consistent across kubelet restarts and therefore
across the watch calls where we watch for changes to the unix socket so we make
this part of Stub object instead of a parameter.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: sample-dp: set default value for `pluginSocksDir`

Setting a reasonable default in case `PLUGIN_SOCK_DIR`
environment variable is not specified.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-e2e-gce

unrelated test failure

swatisehgal commented:
/test pull-kubernetes-node-e2e-containerd

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -84,12 +96,15 @@ func NewDevicePluginStub(devs []*pluginapi.Device, socket string, name string, p
 
 		allocFunc:             defaultAllocFunc,
 		getPreferredAllocFunc: defaultGetPreferredAllocFunc,
+		termSignal:            make(chan bool, 1),
```
I see you writing to this channel, but I don't see it being consumed anywhere.

swatisehgal replied:
Thanks for catching this. I had added this with the understanding that `termSignal` would be needed for handling termination within ListAndWatch() but later realized we already have `stop` channel for that and missed removing this. I will update.

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -53,9 +53,13 @@ type Stub struct {
 	registrationStatus chan watcherapi.RegistrationStatus // for testing
 	endpoint           string                             // for testing
 
+	stopWatcher chan bool
```
Can you use a chan struct{} and close it instead of pushing `true` to it? That is standard practice for "signal" style channels

swatisehgal replied:
Sure, will update.

swatisehgal replied:
This was no longer needed when an fsnotify watcher was added in the stub.

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -159,6 +194,37 @@ func (m *Stub) Stop() error {
+func (m *Stub) Watch(kubeletEndpoint, resourceName, pluginSockDir string) {
+	for {
+		select {
+		case stop := <-m.stopWatcher:
+			if stop {
+				return
+			}
```
With my suggestion above, this would turn into:
```
		case <-m.stopWatcher:
			return
```

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -144,13 +163,29 @@ func (m *Stub) Start() error {
 		return nil
 	}
+
+	m.stopWatcher <- true
```
Likewise, this would turn into `close(m.stopWatcher)`

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -159,6 +194,37 @@ func (m *Stub) Stop() error {
 	return m.cleanup()
 }
 
+func (m *Stub) Watch(kubeletEndpoint, resourceName, pluginSockDir string) {
```
Why do we need our own custom Watch() function rather than just adding the KubeletSocket file to the fsNotify that we already have?

For example, we handle reregistration gracefully in the NVIDIA plugin like so:
https://github.com/NVIDIA/k8s-device-plugin/blob/main/cmd/nvidia-device-plugin/main.go#L155

swatisehgal replied:
Thanks for pointing me to the Nvidia plugin. Sure, using fsnotify seems like a better approach. I will explore that and make updates.

swatisehgal replied:
I have made updates based on your suggestion and captured that as a separate commit in case we need to iterate over it.

swatisehgal committed:
node: sample-dp: Use fsnotify for kubelet restart detection

Add kubeletSocket file to fsnotify instead of polling and waiting for deletion
of device plugin unix socket as a way of detecting kubelet restart. We need to
ensure that the device plugin re-registers itself after kubelet restart depending
on the configured registration mode (auto-registration or controller registration).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
@klueska @SergeyKanzhelev A gentle ping: could you please take a look at this PR.

SergeyKanzhelev commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -76,20 +81,36 @@ func defaultAllocFunc(r *pluginapi.AllocateRequest, devs map[string]pluginapi.De
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.ErrorS(err, "Watcher creation failed")
+		panic(err)
```
how do we decide between `panic` and returning error? Is the kubelet watcher a critical functionality?

swatisehgal replied:
The watcher here watches for changes to a directory where all the device plugin sockets are placed which is important for us to monitor to identify kubelet restarts. When a new instance of kubelet starts, it deletes all Unix sockets under the directory being watched here (/var/lib/kubelet/device-plugins). In case the watcher fails it seems critical enough to me to surface this error. I chose panic here because the alternative here (like you suggested) is to return an error but that would impact device plugin implementations that are using this `NewDevicePluginStub` to bootstrap their respective plugins. I am aware of at least [one device plugin implementation](https://github.com/k8stopologyawareschedwg/sample-device-plugin/blob/main/pkg/deviceplugin/deviceplugin.go) that relies on this and didn't want these plugins to break in case they upgrade to Kubernetes latest toolchain.

If you feel strongly about it, please let me know and I will make updates to return an error here instead and let the plugins handle the error themselves.

SergeyKanzhelev replied:
sorry, got interrupted on this PR review and then lost track of it. Coming back here.

So you are saying that without the kubelet restart monitoring functionality this stub doesn't make sense to use? If so, `panic` is OK

swatisehgal replied:
Yep, pretty much. A device  plugin is expected to handle kubelet restarts as has been documented [here](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#handling-kubelet-restarts).

SergeyKanzhelev commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -159,6 +208,35 @@ func (m *Stub) Stop() error {
+				klog.InfoS("inotify: file created, restarting", "kubeletEndpoint", kubeletEndpoint)
+				if err := m.Restart(); err != nil {
+					klog.ErrorS(err, "Unable to restart server")
+					panic(err)
```
should we retry maybe couple of times. I am thinking of scenario when kubelet just created the socket, but not serving yet. Or some other race like this. I had bad experience with file watchers on Windows where notificaiton come very early (not this socket, but in general) and not sure how they work on Linux.

swatisehgal replied:
Sure, that seems like a reasonable concern. I will add a mechanism to retry. Thanks!

SergeyKanzhelev commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -76,20 +81,36 @@ func defaultAllocFunc(r *pluginapi.AllocateRequest, devs map[string]pluginapi.De
 }
 
+// stubRegisterControlFunc is the function called when a registration request is received from Kubelet
+type stubRegisterControlFunc func() bool
```
so this Stub, is it only for the sample plugin? Or we expect more prod usage?

swatisehgal replied:
`stubRegisterControlFunc` is referenced in the `Stub` struct above which can be used by other device plugins for their bootstapping with the help of `NewDevicePluginStub` initializer. Sample device plugin happens to be such implementation that we use for testing.

swatisehgal committed:
node: sample-dp: Add retry to handle device plugin restart failure

Add retry mechanism to handle cases where after kubelet restarts, the device
plugin unix socket(s) were created but not ready to serve yet.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

swatisehgal commented:
/assign @klueska @dchen1107 
for approval on container manager updates
/assign @dims
for  approval on`test/images` updates

dims commented:
/lgtm
/approve

dims commented:
/hold for other reviewers as @swatisehgal mentioned

swatisehgal commented:
LGTM label was removed as the PR had to be rebased.

SergeyKanzhelev commented:
/lgtm

tedyu commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go:
```diff
@@ -159,6 +208,46 @@ func (m *Stub) Stop() error {
+					lastErr = restartErr
+					return false, nil
+				})
+				if err != nil {
```
the two `return` statements above return `nil` as error.

when would this `err` be non-nil ?

swatisehgal replied:
`PollUntilContextTimeout` polls until it returns true, an error, or hits a deadline.

The aim here is to allow time (at least 2 minute) for device plugin to restart. Please refer to this [ review comment](https://github.com/kubernetes/kubernetes/pull/118534#discussion_r1331931164) for the reason for adding the retry mechanism. We want to stop polling only if the plugin restart succeeds (and therefore exit with `true` and `nil` error). If the restart fails (or for some other reason) we return a non-nil error the polling would stop and we won't retry. Because we want to poll until the restart happens successfully and timeout only after 2 minutes if restart does not succeed we don't return a non-nil error.

derekwaynecarr commented:
/approve

swatisehgal commented:
/unhold
